drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -182,144 +222,161 @@ static int bq274xx_gauge_configure(const struct device *dev)
+	checksum ^= BQ27427_CC_GAIN_SIGN_BIT;
 
-	reg_val = sys_cpu_to_be16(designenergy_mwh);
-	ret = i2c_burst_write_dt(&config->i2c,
-				 BQ274XX_EXT_BLKDAT(regs->dm_design_energy),
-				 (uint8_t *)&reg_val, sizeof(reg_val));
-	if (ret < 0) {
-		LOG_ERR("Failed to write design energy");
-		return -EIO;
-	}
+	LOG_DBG("bq27427: val=%02x checksum=%02x", val, checksum);
 
-	reg_val = sys_cpu_to_be16(config->terminate_voltage);
-	ret = i2c_burst_write_dt(&config->i2c,
-				 BQ274XX_EXT_BLKDAT(regs->dm_terminate_voltage),
-				 (uint8_t *)&reg_val, sizeof(reg_val));
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ27427_CC_GAIN, val);
 	if (ret < 0) {
-		LOG_ERR("Failed to write terminate voltage");
+		LOG_ERR("Failed to update ccgain");
 		return -EIO;
 	}
 
-	reg_val = sys_cpu_to_be16(taperrate);
-	ret = i2c_burst_write_dt(&config->i2c,
-				 BQ274XX_EXT_BLKDAT(regs->dm_taper_rate),
-				 (uint8_t *)&reg_val, sizeof(reg_val));
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_CHECKSUM, checksum);
 	if (ret < 0) {
-		LOG_ERR("Failed to write taper rate");
+		LOG_ERR("Failed to update block checksum");
 		return -EIO;
 	}
 
-	ret = bq274xx_read_data_block(dev, 0, block, sizeof(block));
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	return 0;
+}
+
+static int bq274xx_gauge_configure(const struct device *dev)
+{
+	const struct bq274xx_config *const config = dev->config;
+	struct bq274xx_data *data = dev->data;
+	const struct bq274xx_regs *regs = data->regs;
+	int ret;
+	uint16_t designenergy_mwh, taperrate;
+	uint8_t block[BQ27XXX_DM_SZ];
+	bool block_modified = false;
+
+	designenergy_mwh = (uint32_t)config->design_capacity * 37 / 10; /* x3.7 */
+	taperrate = config->design_capacity * 10 / config->taper_current;
+
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
 	if (ret < 0) {
-		LOG_ERR("Unable to read block data");
+		LOG_ERR("Unable to unseal the battery");
 		return -EIO;
 	}
 
-	checksum_new = 0;
-	for (uint8_t i = 0; i < ARRAY_SIZE(block); i++) {
-		checksum_new += block[i];
-	}
-	checksum_new = 255 - checksum_new;
-
-	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_CHECKSUM, checksum_new);
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_B);
 	if (ret < 0) {
-		LOG_ERR("Failed to update new checksum");
+		LOG_ERR("Unable to unseal the battery");
 		return -EIO;
 	}
 
-	tmp_checksum = 0;
-	ret = i2c_reg_read_byte_dt(&config->i2c, BQ274XX_EXT_CHECKSUM, &tmp_checksum);
+	ret = bq274xx_change_mode(dev, BQ274XX_CTRL_SET_CFGUPDATE);
 	if (ret < 0) {
-		LOG_ERR("Failed to read checksum");
-		return -EIO;
+		return ret;
 	}
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_BAT_INSERT);
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_CONTROL, 0x00);
 	if (ret < 0) {
-		LOG_ERR("Unable to configure BAT Detect");
+		LOG_ERR("Failed to enable block data memory");
 		return -EIO;
 	}
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_SOFT_RESET);
+	ret = bq274xx_read_block(dev, 82, block, sizeof(block));
```

Comment:
A define for 82 would be good

Reply:
Yeah, there's no name for it though, but yeah the magic looks off, I added a couple defines for the subclasses, at least they are explicit now.